### ISSUE TITLE
FeatSeekR

### ISSUE THREAD

**Issue by tcapraz** [Bioconductor/Contributions#2945]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/tcapraz/FeatSeekR

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand Bioconductor [Package Naming Policy][9] and acknowledge
  Bioconductor may retain use of package name.

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.
  
- [x] I am familiar with the [Bioconductor code of conduct][7] and 
  agree to abide by it.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For questions/help about the submission process, including questions about
the output of the automatic reports generated by the SPB (Single Package
Builder), please use the #package-submission channel of our Community Slack.
Follow the link on the home page of the [Bioconductor website][8] to sign up.

[1]: https://contributions.bioconductor.org/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/
[7]: https://bioconductor.org/about/code-of-conduct/
[8]: https://bioconductor.org/
[9]: https://bioconductor.org/developers/package-submission/#naming

**Comment by vjcitn** [comment]
Thanks for this submission.  
```
set.seed(111)
# simulate data with 500 samples, 3 replicates and 5 latent factors 
# generating 50 features
samples <- 500
latent_factors <- 5
replicates <- 3

sim <- FeatSeekR::simData(samples=samples,latent_factors =latent_factors,
                replicates = replicates)
data <- sim[[1]]
reps <- sim[[2]]
```
is early in the vignette.  The relationship between features and samples should be made systematic.  In Bioconductor we use SummarizedExperiment to accomplish this.  The contributors guide explains some of the most basic expectations of submissions with respect to interoperation with Bioconductor infrastructure.  I also don't understand how "latent factors" can be described with an integer.  The underlying latent variable models should be made clear, even if the number of latent factors is all that matters, by default.

**Comment by vjcitn** [comment]
Please respond with revisions to the package and commentary here.

**Comment by tcapraz** [comment]
Hi @vjcitn ,

thank you for your comments. The package now makes use of the SummarizedExperiment class. I also added more documentation that explains how the function simData simulates data, I hope that makes it clearer.

Best regards
Tümay

**Comment by HelenaLC** [comment]
First round of comments below! Note that many are suggestions and not strictly required. Please comment back here with what has/not been addressed how/why not, or any comments regarding any needed clarification or questions you might have; happy to help/discuss. Cheers!

# code

- [ ] Consider having a look at the code-style guidelines recommended [here](https://contributions.bioconductor.org/r-code.html#coding-style), e.g., spaces around logical operators and after commas, no spaces around `=`s for function arguments, not exceeding an 80-character limit per line, 4-space tab indentation etc. There are currently a lot of inconsistencies (e.g., 2-3 spaces in some places, one or no spaces around `!=`/`==`, spaces around argument `=`s sometimes but not others etc.).

- [ ] In any function using `message()`, consider adding an argument `verbose=TRUE/FALSE` so that users can control whether or not to print function progress in the console. E.g., `FeatSeek` prints a message for each iteration, which some users might want to suppress.

- [ ] `plotSelectedFeatures()` currently selects assay `selected` internally. This isn't very robust because users can rename assays at any time. I'd suggest to instead expose an argument `assay` (or similar) with default `"selected"`, and extracting the corresponding data via `assay(res, assay)` (which is equivalent to `assays(res)[[assay]]`).

- [ ] You are currently using `package::function` to call external dependencies, and the `NAMESPACE` lists 0 imports. Instead, please use selective import via `roxygen`, i.e., `@importFrom package function1 function2 ...`. Doing this, you needn't (and, except when there are conflicts, shouldn't) use `::`.

- [ ] I feel that overall the `SummarizedExperiment`-related code could be greatly simplified, which is (among many others) why it's desirable using it in the first place... E.g., in *check_input.R* (but analogous in other functions):
  - `fnames <- rownames(SummarizedExperiment::assays(se)$data)` could be `rownames(se)`
  - 3x `SummarizedExperiment::colData(se)$replicates` could be `reps <- se$replicates`
  - `dim(SummarizedExperiment::assays(se)$data)[2]` is equivalent to `ncol(se)`

- [ ] I'd strongly recommend adding some argument validity checks at the top of each function in order to avoid failure downstream and potentially confusing error messages. To give just one (simple) example, in `plotSelectedFeatures()`:

```r
if (is.null(n_features))
  n_features <- nrow(res)
stopifnot(
  is(res, "SummarizedExperiment"),
  # 'n_features' should be a scalar integer and 
  # mustn't exeed the number of available features
  is.numeric(n_features), 
  length(n_features) == 1,
  round(n_features) == n_features,
  # specified 'assay' to use should be a scalar 
  # character that matches a unique assay name
  is.character(assay), length(assay) == 1,
  sum(grepl(assay, assayNames(res))) == 1)
```

# vignette

- [ ] Please refrain from using `@` for accession in the vignette, and anywhere (except for very rare cases, which I think doesn't apply here). There are designated `SummarizedExperiment` accessor and replacement methods. E.g., `sim@colData$replicates` should be `colData(sim)$replicates` or `sim$replicates`.

- [ ] I commented on code-style above. Especially in the vignette, I'd strongly encourage to "clean up" a bit as this will be what users consult most and is typically displayed online in HTML format (e.g., 4-space tab indentation, consistent spacing around argument `=` vs. logical operators, space after `,`s etc.).

- [ ] The vignette currently runs through each function with a one-sentence commentary. It'd be desirable to provide more details here, which is what the vignette is for. E.g., describing function arguments / their effect, methodological details, interpretation of visualizations etc.

- [ ] Perhaps be consistent in using code-style for function, variable, package names. E.g., FeatSeek is a function but written in plain text, dataframe might be better written as `data.frame`.

# other

- [ ] Consider adding a package *.man* page so users would get an overview of the package's basic workflow and key functions with `?FeatSeekR`.

- [ ] In `roxygen` documentation such as `@param` descriptions, consider using `\code{...}` syntax around class, function, variables names etc. to distinguish plane text from R code.

- [ ] In the `DESCRIPTION`, consider adding a `URL:` and `BugReports:` field pointing to the repo code and issues, respectively (i.e., https://github.com/tcapraz/FeatSeekR(/issues)).

- [ ] In the `NEWS` file, please correct the date (currently a placeholder "YYYY-MM-DD") or omit it.

# coverage

- Looking good. However, although the overall coverage is quite high, there are only view hits per line at the moment. So perhaps worth considering globally adding more unit tests.

```r
# devtools::test_coverage()
FeatSeekR Coverage: 84.62%
R/check_input.R: 66.67%
R/plots.R: 70.00%
R/FeatSeekR.R: 88.24%
R/evaluate_features.R: 93.33%
R/fit_lm.R: 100.00%
R/simData.R: 100.00%
```

**Comment by tcapraz** [comment]
Hi @HelenaLC,

thank you very much for your valuable comments! I tried to address all of them.

# code

* changed the code according to the code-style guidelines.
* added a `verbose` argument to `FeatSeek`
* added `assay` argument to `plotSelectedFeatures`
* simplified `SummarizedExperiment` code
* added validity checks in every function
* fixed imports by adding  selective imports (`@importFrom`) to functions

# vignette

* as above, I changed the code according to the code-style guidelines. 
* removed `@` accessor
* fixed notations, where code-style was not used before 
* added more explanatory comments before functions and text between chunks, to better guide the interpretation of the plots

# other

* added a man page for the package (`?FeatSeekR`)
* added `BugReports` url to `DESCRIPTION`
* removed date in `NEWS` file

# coverage

* added more unit tests for every function

**Comment by HelenaLC** [comment]
Hey! Thanks for the quick update. It's come along nicely and I appreciate the effort! I have a couple of very minor comments left, but would encourage spending another small round on this (hopefully < 1h work) to get things into a good shape for release! (*some comments are with good intention, but more for your personal learning / better coding practices.)

- the build report (see above) is telling you that there are portions of the code exceeding 80 characters and/or not obeying to 4-space/tab indentation. Consider browsing through the *.R* scripts and fixing this as much as possible... e.g., line 26+ in *evaluate_features.R* is quite badly formatted. And there is generally still a lot of inconsistency.  
For what it's worth, you can have your RStudio settings do this automatically while coding by setting i) tab width 2 space, ii) vertical alignment off, and iii) display a vertical line at 80 characters for guidance. To speed up things, there is even an "auto-format after pasting" option that would format your code with cmd+A > cmd+C > cmd+V (highlight, copy, paste), though I personally prefer doing it manually if the amount of code is reasonable...

- for readability / easier comprehension (& sticking to an 80-char limit per line), it's nice to unnest code at times (especially in the vignette that is very visible!), e.g., *plots.R* line 84 and vignette line 77: `cor <- stats::cor(t(SummarizedExperiment::assay(sim, "data"))[sim$replicates==1, ], use = "pairwise.complete.obs")`
```
y <- assay(sim, "data")                    # extract
y <- y[sim$replicate == 1, , drop = FALSE] # subset
cor <- cor(t(cor), use = ...)              # correlate
```

- Kindly, please take special care of the above 2 points (style & unnesting) in the vignette! This is what is visible online and badly formatted code (80-char, indentation) will be hard to follow. Might even consider getting rid of `SummarizedExperiment::` syntax (and similar) -- this isn't necessary when importing the package, and makes lines unnecessarily long.

- careful with subsetting matrices. E.g., `cor` expects a matrix as input, but `x[... == 123, ]` will give a vector, if the condition is only met for a single row. `x[..., ..., drop = FALSE]` is meant to prevent this, and will return a matrix no matter what (i.e., a 2D array and not a vector). I'd suggest a global search for `cor(` and fixing this and similar cases that come to mind (i.e., anywhere a matrix is being subset really).

- not sure if that was intentional, but the DESCRIPTION is still missing a `ULR:` field pointing to *https://github.com/tcapraz/FeatSeekR*

- `if (x == TRUE)` can be written as just `if (x)` (e.g., *FeatSeekR.R* line 115)

- consider adding one of the "AssayDomain" and/or "Technology" [BiocViews], e.g., does it apply to any type or sequencing-based data?

**Comment by HelenaLC** [comment]
Just pinging- any update? Shouldn't be too much I think/hope :)

**Comment by tcapraz** [comment]
Hi!

Thank you for the second round of comments. I again tried to address all your comments.

While going over the code, I noticed it would be good to reimplement the function `calcFstat` in FeatSeekR.R, to make it easier to understand and more in line with the manuscript I'm currently writing. 
The new implementation should return equivalent results and is a bit faster.

Thanks!

**Comment by HelenaLC** [comment]
- The build report gives a NOTE stating that the LICENSE is not mentioned in the DESCRIPTION file.
- In the vignette, it would be nice to direct users to external packages via hyperlinks. The `BiocStyle` package has functions to do this automatically, e.g., `Biocpkg("EBImage")`.
- In `FeatSeekR.R` (but perhaps other places?), I noticed `nlevels(fac)`, which will fail if `fac` is not a factor, which is not being checked. Perhaps safest to use `fac <- droplevels(factor(fac))` here, or similar, to assure this does what you want (`droplevels` might or might not make sense, depending on whether or not missing levels are being considered somehow).
- Generally, I'm seeing lots of (nested) `for`-loops, which are rarely necessary and efficient in R... I'd highly encourage getting familiar with `l/s/v/m/apply` for future development/as a programmer in general. Sometimes, just base R will also do. Specially initializing an empty object and appending to it is very bad practice (see [Vectorize](https://contributions.bioconductor.org/r-code.html#vectorize) for more details)... E.g.,  `simData.R` lines  85+ should most definitely use `lapply` to construct `replist`. And 100+
```
featnames <- c()
for (i in seq_len(n_latent_factors)) {
  featnames <- c(featnames,
    paste0("Latent_factor", i, "_Feature", seq_len(10))
  )
}
# can be written as
lfac <- sprintf("Latent_factor%s", seq_len(n_latent_factors))
feat <- paste0("_Feature", seq_len(n))
featnames <- paste0(rep(lfac, each = n), feat)
```

**Comment by lshep** [comment]
The default branch of your GitHub repository has been added to Bioconductor's
git repository as branch devel.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/tcapraz.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("FeatSeekR")`. The package 'landing page' will be created at

https://bioconductor.org/packages/FeatSeekR

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
